Introducing CVCUDA Backend

AntoineSimoulin opened the pull request:
Summary
-------

This PR provides the first building blocks for CV-CUDA integration in torchvision. We add the functionals `to_cvcuda_tensor` and `cvcuda_to_tensor` to transform from `torch.Tensor` to `cvcuda.Tensor` and back. We also implement the corresponding class transforms `ToCVCUDATensor` and `CVCUDAToTensor`. These transforms require CV-CUDA to be installed.

How to use
----------

```python
from PIL import Image
import torchvision.transforms.v2.functional as F 

# Create rand ``torch.Tensor`` image (must be 3-channel RGB/Gray and have batch dimension)
img_tensor = torch.randint(0, 256, (1, 3, 320, 240), dtype=torch.uint8)

# Convert to ``cvcuda.Tensor`` (will be uploaded to CUDA)
cvcuda_tensor = F.to_cvcuda_tensor(img_tensor)

# Convert back to ``torch.Tensor``
img_tensor = F.cvcuda_to_tensor(cvcuda_tensor)
```

> [!NOTE]
> 
> *   ``cvcuda.Tensor``  are automatically converted to NHWC shape (since most CV-CUDA transforms only support this shape)
> *   Only 3-channel RGB images and 1-channel grayscale are supported for now
> *   Input tensors will be uploaded to CUDA device when converting to CV-CUDA tensors
> *   CV-CUDA must be installed: `pip install cvcuda-cu12` (CUDA 12) or `pip install cvcuda-cu11` (CUDA 11)

Run unit tests
--------------

```bash
pytest test/test_transforms_v2.py -k "cvcuda"
...
338 passed, 9774 deselected in 1.65s
```

Differential Revision: D85862362



cc @vfdev-5

AntoineSimoulin commented:
NOTES:
* Should we upload tensors `.to(cuda)` directly inside `to_nvcv_tensor` function?
* Should we keep CVCUDA specific functions inside isolated "_cvcuda" files or within existing files (e.g. "_geometry")?
* NVCV tensors are automatically converted to NHWC format. Contrary to torchvision convention, which relies on NCHW format. If we call `to_nvcv_tensor` and then `nvcv_to_tensor`, we might have an output tensor with a different shape.
* Are the image format from `_infer_nvcv_format` exhaustive and aligned with what is done for PIL?
* What is the difference between `nvcv.Image` and `nvcv.Tensor`?

NicolasHug commented on torchvision/transforms/v2/functional/_cvcuda.py:
```diff
@@ -0,0 +1,243 @@
+import warnings
```
We discussed a bit offline already, but I'd prefer sticking to our current structure where all related kernels live in the same place.  It's makes it easier to compare kernel implementations for all the backends we support, and preserves consistency of the file structure. That is, I think these kernels and functionals should be in `torchvision/transforms/v2/functional/_type_conversion.py` . I'm hoping that each kernel can just lazy-import nvcv - that's cheap, imports are cached. 


We might need the following workarounds:

- for type annotation, use the following pattern (from `claude`):

```
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    import lib

def f(some_param: "lib.some_class") -> None:
    # Your implementation
    pass
```

I think you had something like that before?


- for dispatcher registration, we'll probably still need to create this `_cvcuda.py` file. But the implementation of the kernel itself could still live in the other files like `torchvision/transforms/v2/functional/_type_conversion.py`. This file would only contain the kernel **registration**.

NicolasHug commented on torchvision/transforms/v2/functional/_type_conversion.py:
```diff
@@ -25,3 +29,176 @@ def to_image(inpt: Union[torch.Tensor, PIL.Image.Image, np.ndarray]) -> tv_tenso
+        TypeError: If dtype is not supported for the given number of channels.
+        ValueError: If number of channels is not 1 or 3.
+    """
+    import nvcv  # type: ignore[import-not-found]
```
I think for all of these that happen within kernels, we'll want to use a helper that raises an informative error message if nvcv isn't installed. Something like

```py
def import_nvcv():
    try:
        import nvcv
        return nvcv
    except ImportError as e:
        raise RuntimeError("You need nvcv see blahblahblah for installing it") from e

# inside of kernel / functional:
nvcv = import_nvcv()
```

NicolasHug commented on torchvision/transforms/v2/functional/_type_conversion.py:
```diff
@@ -25,3 +29,176 @@ def to_image(inpt: Union[torch.Tensor, PIL.Image.Image, np.ndarray]) -> tv_tenso
+        elif dtype == torch.float32:
+            return nvcv.Format.RGBf32, img_tensor
+        else:
+            raise TypeError(f"Unsupported dtype {dtype} for 3-channel image")
```
Do I understand correctly that torch.float64 isn't supported by CV-CUDA? That's fine, just checking.

NicolasHug commented on torchvision/transforms/v2/functional/_type_conversion.py:
```diff
@@ -25,3 +29,176 @@ def to_image(inpt: Union[torch.Tensor, PIL.Image.Image, np.ndarray]) -> tv_tenso
+    if img_tensor.ndim not in (3, 4):
+        raise ValueError(f"pic should be 2/3/4 dimensional. Got {img_tensor.ndim} dimensions.")
+
+    # For batched inputs, use the first image to infer format
```
Below: just add a fake batch dim to the 3D input so that it becomes 4D. Then you can assume it's always 4D and avoid all the if/else checks.

Although I am almost tempted to not support 3D inputs. CV-CUDA will mostly be faster for large batch sizes, so supporting 3D unbatched data adds little value. We could add it later if needed and keep things simpler for now. Thoughts?

justincdavis commented on torchvision/transforms/v2/functional/_type_conversion.py:
```diff
@@ -25,3 +52,134 @@ def to_image(inpt: Union[torch.Tensor, PIL.Image.Image, np.ndarray]) -> tv_tenso
+
+    # Convert NVCV Tensor to PyTorch tensor via CUDA array interface
+    # NVCV tensors expose __cuda_array_interface__ which PyTorch can consume directly
+    cuda_tensor = torch.as_tensor(nvcv_img.cuda(), device="cuda")
```
Yes, the cvcuda.Tensor object does not expose cuda_array_interface, the .cuda() method returns a cvcuda.ExternalBuffer object with the attribute.

AntoineSimoulin commented on test/test_transforms_v2.py:
```diff
@@ -6733,6 +6742,184 @@ def test_functional_error(self):
+            F.to_cvcuda_tensor(img_data)
+
+    @pytest.mark.parametrize("num_channels", [1, 3])
+    @pytest.mark.parametrize("dtype", [torch.uint8, torch.float32, torch.float64])
```
Gotcha! I think at this point we should remove the `_validate_cvcuda_dtype` function since cvcuda should support all data types supported by torchvision (`torch.uint8`, `torch.int16`, `torch.int32`, `torch.float32`, `torch.float64`). Maybe just reject `torch.float16`?

AntoineSimoulin replied:
For the following PRs, yes I think it would be useful to have a list of the datatypes/channels CV-CUDA operators can utilize!

justincdavis commented on test/test_transforms_v2.py:
```diff
@@ -6733,6 +6742,125 @@ def test_functional_error(self):
+            img_data = img_data.cuda()
+            F.to_cvcuda_tensor(img_data)
+
+    @pytest.mark.parametrize("num_channels", [1, 3])
```
CV-CUDA can support 2/4-channel tensors, I missed this previously when discussing the validate format function. Same comment for `test_round_trip_batched`.

justincdavis commented on torchvision/transforms/v2/functional/_type_conversion.py:
```diff
@@ -25,3 +51,82 @@ def to_image(inpt: Union[torch.Tensor, PIL.Image.Image, np.ndarray]) -> tv_tenso
+    Args:
+        pic (torch.Tensor): Image to be converted to cvcuda.Tensor.
+            Tensor can be in CHW format (unbatched) or NCHW format (batched).
+            Only 1-channel and 3-channel images are supported.
```
Should update this comment to include 2/4 channel images since we no longer filter those out with the validation function.

AntoineSimoulin replied:
@justincdavis thanks for the comments. See my answer above!

AntoineSimoulin commented:
Summary of the changes:
* Cleaned up docstrings for `ToCVCUDATensor` and `CVCUDAToTensor`
* Moved `_import_cvcuda` to "torchvision/transforms/v2/functional/_utils" since it is used in multiple files
* Update `to_cvcuda_tensor` and `cvcuda_to_tensor` to only support batched NHWC layout as input (no single image with HWC layout) to simplify the logic from this initial commit
* Cleaned up docstrings for `to_cvcuda_tensor` and `cvcuda_to_tensor` 
* Improved test parametrization in "test/transforms_v2"
* Added kernel for `get_size_image_cvcuda` to be used in tests

AntoineSimoulin commented:
@justincdavis thanks for the confirmation, just wanted to double check and indeed I confirmed `isinstance` works without issue.

justincdavis commented:
Other than the final comment I made, this looks great to me! Thank you @AntoineSimoulin !

NicolasHug commented on torchvision/transforms/v2/_type_conversion.py:
```diff
@@ -90,3 +93,48 @@ class ToPureTensor(Transform):
+
+
+class CVCUDAToTensor(Transform):
+    """Convert a `cvcuda.Tensor` with NHWC layout to a `torch.Tensor` with NCHW layout.
```
```suggestion
    """Convert a `cvcuda.Tensor` with  to a `torch.Tensor` with NCHW shape.
```

AntoineSimoulin commented on torchvision/transforms/v2/_type_conversion.py:
```diff
@@ -90,3 +93,48 @@ class ToPureTensor(Transform):
+        >>> img_tensor = torch.randint(0, 256, (1, 3, 320, 240), dtype=torch.uint8)
+        >>> img_cvcuda = v2.ToCVCUDATensor()(img_tensor)
+        >>> print(img_cvcuda.shape)
+        (1, 3, 240, 320)
```
Ok let me remove the examples.
